### ISSUE TITLE
Has.Count does not work allways

### ISSUE THREAD

**Issue by NUnitTester** [nunit/nunit#1816]
If the collection is an Array the `Has.Length` does work but the `Has.Count` doesn't. That's wrong because every array is also a collection which has a property Count.

`int[] numbers = {1, 2, 3};`
`Assert.That(numbers, Has.Length.EqualTo(3));`
`Assert.That(numbers, Has.Count.EqualTo(3));`
`ICollection<int> numbersCollection = numbers;`
`Assert.That(numbersCollection, Has.Count.EqualTo(3));`

**System.ArgumentException : Property Count was not found**

**Comment by CharliePoole** [comment]
This is by design. We only look at properties based on the type passed, just as the C# compiler does. You would have to cast numbers to some type to IEnumerable in order for the test to pass.

I recognize that there are arguments for implementing it as you describe. We considered those arguments when the feature was added some years back but decided to limit ourselves to properties of the defined Type.

**Comment by rprouse** [comment]
@NUnitTester you say that the array has a property Count, but that isn't technically correct. LINQ provides an extension method `Count()`. 

I can understand how you would expect `Has.Count` to work, but as @CharliePoole said, we've already made the choice of where to draw the line. Supporting `Count` would probably require us to attempt a cast to an `IEnumerable<T>` then apply the `Count()` extension method. This is extra complexity that we decided against.

**Comment by appel1** [comment]
Arrays in .net implements ICollection that has a Count property. It is explicitly implemented though so you'd have to cast to ICollection or take explicit implementations into account when reflecting. Perhaps that'd be confusing though.

**Comment by ChrisMaddock** [comment]
This looks like #1394

**Comment by CharliePoole** [comment]
@ChrisMaddock I guess you are right, although I initially took them differently.

My take on #1394 was that we don't __even__ see the property when you pass in the interface that is implemented explicitly. That's not actually what @yaakov-h 's example shows there, but it was how I read it at the time.

It seems to me correct that you should not see an explicitly implemented property if you pass in the object without casting (this issue) but that you should see it when passing in the object cast to the interface. Currently you don't see it in either case.

Let's decide two things here: (1) how it should be and (2) how it can be right now, since our assertions are currently based on objects. Then we should close this issue as a duplicate of #1394.

**Comment by NUnitTester** [comment]
First thing: It would be nice if the item is closed after has had a change to react.

@rprouse: As mentioned I haven't used LINQ at all and Count() is a function not a property. You will see this part of code I used:
`int[] numbers = {1, 2, 3};`
`ICollection<int> numbersCollection = numbers;`
`Assert.That(numbersCollection, Has.Count.EqualTo(3));`

**System.ArgumentException : Property Count was not found**

Here the Assert get an ICollection<int> and does not work correctly. The interface give back an ICollection<?> and it depends on implementation what really is used. I cannot know it, can I.
So if ICollection<?> is used, Count is expected to work and implemented. And it does not.

@CharliePoole : To answer your questions ...
(1) I think, you have described the problem: it goes  only on the implementation class not on the implemented interfaces. In my case, I cannot see the implementation, I only got a implemented interface which has a property Count.
(2) I understand that if one use the implementation class Count is not found. But as far as I know is there a possibility to search in derived interfaces also per reflection in the same method call. This should work easily (but not sure).

**Comment by NUnitTester** [comment]
@ChrisMaddock : Thanks for the link. Has I have found out reading the discussion above, this go deeper in Nunit as I have thought.

**Comment by NUnitTester** [comment]
`ICollection<int> numbersCollection = IWorker.DoMethod();`
`Assert.That(numbersCollection, Has.Count.EqualTo(3));`

Think about the following usecase: You will test different implementations of IWorker given as parameter of tha attribute TestFixture, in my case 3. Two will use an Array to give back as ICollection<int> and one will give back an HashSet<int>. 
How to write the Unit-Test with Has.Count or Has.Length?

And don't answer to use Is.EqualTo ... I know it.

**Comment by rprouse** [comment]
@NUnitTester we are always able to reopen issues and as you see, it is fine to continue discussion on closed issues to try to convince us that we made the wrong decision. This is issue #1816, NUnit is a very popular project with a large volume of issues reported and very few people to respond to the issues, so we do our best to deal with issues quickly and efficiently.

Your example above is more compelling, especially if `IWorker.DoMethod()` returns an `ICollection<int>`. That said, we also need to think about the internal implementation within NUnit and the complexity & performance issues it might introduce.

As @CharliePoole says, this is a duplicate of #1394, so we will leave this closed. I will copy your use case across to that issue.

**Comment by CharliePoole** [comment]
@rprouse I still think we have two things to decide before we can work on #1394. I'll move my above comment to that issue.
